appsec/src/extension/backtrace.c

```diff
@@ -0,0 +1,280 @@
+
+            zend_hash_next_index_insert_new(datadog_backtrace_ht, &new_frame);
+            if (--top == 0) {
+                break;
+            }
+        }
+        ZEND_HASH_FOREACH_END();
+    }
+
+    if (bottom > 0) {
+        unsigned int position = frames_on_stack - bottom;
+        DD_FOREACH_FROM(php_backtrace_ht, 0, position, index)
+        {
+            tmp = _z;
+            zval new_frame;
+
+            if (!php_backtrace_frame_to_datadog_backtrace_frame(
+                    tmp, &new_frame, index)) {
+                continue;
+            }
+
+            zend_hash_next_index_insert_new(datadog_backtrace_ht, &new_frame);
+        }
+        ZEND_HASH_FOREACH_END();
+    }
+}
+
+void generate_backtrace(zend_string *id, zval *dd_backtrace)
+{
+    array_init(dd_backtrace);
+
+    if (!get_global_DD_APPSEC_STACK_TRACE_ENABLED() || !id) {
+        return;
+    }
+
+    zval language;
+    ZVAL_STR_COPY(&language, _php_value);
+    zval id_zv;
+    ZVAL_STR_COPY(&id_zv, id);
+    zend_hash_add(Z_ARRVAL_P(dd_backtrace), _language_key, &language);
+    zend_hash_add(Z_ARRVAL_P(dd_backtrace), _id_key, &id_zv);
+
+    zval frames;
+    zval php_backtrace;
+    zend_fetch_debug_backtrace(
+        &php_backtrace, 1, DEBUG_BACKTRACE_IGNORE_ARGS, NO_LIMIT);
+    php_backtrace_to_datadog_backtrace(&php_backtrace, &frames);
+    zend_hash_add(Z_ARRVAL_P(dd_backtrace), _frames_key, &frames);
+
+    zval_dtor(&php_backtrace);
+}
+
+static PHP_FUNCTION(datadog_appsec_testing_generate_backtrace)
+{
+    zend_string *id = NULL;
+    if (zend_parse_parameters(ZEND_NUM_ARGS(), "S", &id) != SUCCESS) {
+        RETURN_FALSE;
+    }
+
+    generate_backtrace(id, return_value);
+}
+
+zval *dd_hash_find_or_new(HashTable *ht, zend_string *key)
+{
+    zval *result = zend_hash_find(ht, key);
+
+    if (!result) {
+        zval new_zv;
+        result = zend_hash_add(ht, key, &new_zv);
+    }
+
+    return result;
+}
+
+bool report_backtrace(zend_string *id)
+{
+    zend_object *span = dd_trace_get_active_root_span();
+    if (!span) {
+        if (!get_global_DD_APPSEC_TESTING()) {
+            mlog(dd_log_warning, "Failed to retrieve root span");
+        }
+        return false;
+    }
+
+    zval *meta_struct = dd_trace_span_get_meta_struct(span);
+    if (!meta_struct) {
+        if (!get_global_DD_APPSEC_TESTING()) {
+            mlog(dd_log_warning, "Failed to retrieve root span meta_struct");
+        }
+        return false;
+    }
+
+    if (Z_TYPE_P(meta_struct) != IS_ARRAY) {
+        array_init(meta_struct);
+    }
+
+    zval *dd_stack = dd_hash_find_or_new(Z_ARR_P(meta_struct), _dd_stack_key);
+    if (Z_TYPE_P(dd_stack) != IS_ARRAY) {
+        array_init(dd_stack);
+    }
+
+    zval *exploit = dd_hash_find_or_new(Z_ARR_P(dd_stack), _exploit_key);
+    if (Z_TYPE_P(exploit) != IS_ARRAY) {
+        array_init(exploit);
+    }
+
+    if (zend_array_count(Z_ARR_P(exploit)) ==
+        get_global_DD_APPSEC_MAX_STACK_TRACES()) {
```

Comment:
Maybe a debug log statement could be useful here.

Reply:
Also `DD_APPSEC_MAX_STACK_TRACES` could be 0, in which case we aren't supposed to limit the number of stack traces.